gc/gc-iceberg-files/src/main/java/org/projectnessie/gc/iceberg/files/IcebergFiles.java

```diff
@@ -0,0 +1,259 @@
+import org.apache.hadoop.fs.RemoteIterator;
+import org.apache.iceberg.aws.s3.S3FileIO;
+import org.apache.iceberg.io.BulkDeletionFailureException;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.io.ResolvingFileIO;
+import org.immutables.value.Value;
+import org.projectnessie.gc.files.DeleteResult;
+import org.projectnessie.gc.files.DeleteSummary;
+import org.projectnessie.gc.files.FileDeleter;
+import org.projectnessie.gc.files.FileReference;
+import org.projectnessie.gc.files.FilesLister;
+import org.projectnessie.gc.files.NessieFileIOException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Provides functionality to {@link FilesLister list} and {@link FileDeleter delete} files using
+ * Iceberg's {@link S3FileIO} for S3 schemes and/or {@link ResolvingFileIO} for non-S3 schemes.
+ *
+ * <p>The {@link FileIO} instances are only instantiated when needed.
+ */
+@Value.Immutable
+public abstract class IcebergFiles implements FilesLister, FileDeleter, AutoCloseable {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(IcebergFiles.class);
+
+  public static Builder builder() {
+    return ImmutableIcebergFiles.builder();
+  }
+
+  static Stream<String> filesAsStrings(Stream<FileReference> fileObjects) {
+    return fileObjects.map(FileReference::absolutePath).map(URI::toString);
+  }
+
+  static URI ensureTrailingSlash(URI uri) {
+    if (uri.getPath().endsWith("/")) {
+      return uri;
+    }
+    return URI.create(uri + "/");
+  }
+
+  public interface Builder {
+    Builder hadoopConfiguration(Configuration hadoopConfiguration);
+
+    Builder putProperties(String key, String value);
+
+    Builder putAllProperties(Map<String, ? extends String> entries);
+
+    Builder properties(Map<String, ? extends String> entries);
+
+    IcebergFiles build();
+  }
+
+  @Value.Default
+  Configuration hadoopConfiguration() {
+    return new Configuration();
+  }
+
+  abstract Map<String, String> properties();
+
+  @SuppressWarnings("immutables:incompat")
+  private volatile boolean hasResolvingFileIO;
+
+  @SuppressWarnings("immutables:incompat")
+  private volatile boolean hasS3FileIO;
+
+  @Value.Lazy
+  public FileIO resolvingFileIO() {
+    ResolvingFileIO fileIO = new ResolvingFileIO();
+    fileIO.initialize(properties());
+    fileIO.setConf(hadoopConfiguration());
+    hasResolvingFileIO = true;
+    LOGGER.debug("Instantiated Iceberg's ResolvingFileIO");
+    return fileIO;
+  }
+
+  @Value.Lazy
+  S3FileIO s3() {
+    S3FileIO fileIO = new S3FileIO();
+    fileIO.initialize(properties());
+    hasS3FileIO = true;
+    LOGGER.debug("Instantiated Iceberg's S3FileIO");
+    return fileIO;
+  }
+
+  @Override
+  public void close() {
+    try {
+      if (hasS3FileIO) {
+        s3().close();
+      }
+    } finally {
+      if (hasResolvingFileIO) {
+        resolvingFileIO().close();
+      }
+    }
+  }
+
+  private boolean isS3(URI uri) {
+    switch (uri.getScheme()) {
```

Comment:
I have found two issues in my local testing today. (spark with local file system, **file scheme is not present**)

a. NPE here when local filesystem is used without a scheme.
b. Expire is cleaning up whole table. (reason is while adding to bloom filter scheme is not present. But while checking against bloom filter scheme is present and it is not relative path)

```
2022-10-04 16:19:31,254 [ForkJoinPool-2-worker-3] ERROR o.p.g.expire.PerContentDeleteExpired - Adding to bloom filter data/00000-0-6799bf60-588c-473a-b506-88ab137aa006-00001.parquet
2022-10-04 16:19:31,316 [ForkJoinPool-2-worker-3] ERROR o.p.g.expire.PerContentDeleteExpired - Checking against bloom filter file:///tmp/nessie2/db1/t1/data/00000-0-6799bf60-588c-473a-b506-88ab137aa006-00001.parquet
```

Reply:
More context / reproducer would really help here.

Reply:
steps:

./gradlew quarkusDev


/Users/ajantha/Downloads/spark-3.3.0-bin-hadoop2/bin/spark-sql --jars /Users/ajantha/Downloads/iceberg-spark-runtime-3.3_2.12-0.14.1.jar,/Users/ajantha/Downloads/nessie-spark-extensions-3.3_2.12-0.43.1-SNAPSHOT.jar --conf spark.sql.extensions=org.apache.iceberg.spark.extensions.IcebergSparkSessionExtensions,org.projectnessie.spark.extensions.NessieSparkSessionExtensions  --conf spark.sql.catalog.nessie.uri=http://localhost:19120/api/v1 --conf spark.sql.catalog.nessie.ref=main --conf spark.sql.catalog.nessie.authentication.type=NONE --conf spark.sql.catalog.nessie.catalog-impl=org.apache.iceberg.nessie.NessieCatalog --conf spark.sql.catalog.nessie.warehouse=$PWD/nessie2 --conf spark.sql.catalog.nessie=org.apache.iceberg.spark.SparkCatalog


execute from the spark sql --> 
create table nessie.db1.t1(id int) using iceberg;
insert into nessie.db1.t1 select 42;
insert into nessie.db1.t1 select 42;
select * from nessie.db1.t1;
describe formatted nessie.db1.t1;

execute from CLI -->
cd /Users/ajantha/Documents/workspace/nessieWorkspace/nessie/gc/gc-tool/build/executable

./nessie-gc gc --inmemory